pkg/controller/awsfederatedaccountaccess/awsfederatedaccountaccess_controller.go

```diff
@@ -0,0 +1,368 @@
+	if r.client.List(context.TODO(), &client.ListOptions{}, federatedRoleList); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Create map to make finding role easier
+	roleMap := make(map[string]awsv1alpha1.AWSFederatedRole)
+	for _, role := range federatedRoleList.Items {
+		roleMap[role.Name] = role
+	}
+
+	// See if requested role exists
+	if _, ok := roleMap[currentFAA.Spec.AWSFederatedRoleName]; !ok {
+		SetStatuswithCondition(currentFAA, "Requested role does not exist", awsv1alpha1.AWSFederatedAccountFailed, awsv1alpha1.AWSFederatedAccountStateFailed)
+		reqLogger.Error(ErrFederatedAccessRoleNotFound, fmt.Sprintf("Resquested role '%s' not found", currentFAA.Spec.AWSFederatedRoleName))
+
+		err := r.client.Status().Update(context.TODO(), currentFAA)
+		if err != nil {
+			reqLogger.Error(err, fmt.Sprintf("Status update for %s failed", currentFAA.Name))
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil
+
+	}
+
+	// Get aws client
+	awsClient, err := awsclient.GetAWSClient(r.client, awsclient.NewAwsClientInput{
+		SecretName: currentFAA.Spec.AwsCredentialSecret.Name,
+		NameSpace:  currentFAA.Spec.AwsCredentialSecret.Namespace,
+		AwsRegion:  "us-east-1",
+	})
+	if err != nil {
+		awsClientErr := fmt.Sprintf("Unable to create aws client for region ")
+		reqLogger.Error(err, awsClientErr)
+		return reconcile.Result{}, err
+	}
+
+	// here create the custom policy in the cluster account
+	err = r.createIAMPolicy(awsClient, roleMap[currentFAA.Spec.AWSFederatedRoleName], *currentFAA)
+	if err != nil {
+		// if we were unable to create the policy fail this CR.
+		SetStatuswithCondition(currentFAA, "Failed to create custom policy", awsv1alpha1.AWSFederatedAccountFailed, awsv1alpha1.AWSFederatedAccountStateFailed)
+		reqLogger.Error(err, fmt.Sprintf("Unable to create policy resquested by '%s'", currentFAA.Name))
+
+		err := r.client.Status().Update(context.TODO(), currentFAA)
+		if err != nil {
+			reqLogger.Error(err, fmt.Sprintf("Status update for %s failed", currentFAA.Name))
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil
+
+	}
+
+	// create role role and apply custom policys and awsmanagedpolicies
+	err = r.createIAMRole(awsClient, roleMap[currentFAA.Spec.AWSFederatedRoleName], *currentFAA)
+	if err != nil {
+		SetStatuswithCondition(currentFAA, "Failed to create role", awsv1alpha1.AWSFederatedAccountFailed, awsv1alpha1.AWSFederatedAccountStateFailed)
+		reqLogger.Error(ErrFederatedAccessRoleFailedCreate, fmt.Sprintf("Unable to create role requested by '%s'", currentFAA.Name), "AWS ERROR: ", err)
+
+		err := r.client.Status().Update(context.TODO(), currentFAA)
+		if err != nil {
+			reqLogger.Error(err, fmt.Sprintf("Status update for %s failed", currentFAA.Name))
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	// Get account number of cluster account
+	gciOut, err := awsClient.GetCallerIdentity(&sts.GetCallerIdentityInput{})
```

Comment:
Same question of API failure

Reply:
We have the account CR where we could retrieve this from instead of calling AWS on every reconcile loop. OCM would have to provide the Account CR or Claim name.

Reply:
If we did that we would also not need the secret name.

Reply:
:+1: